x/stakeibc/keeper/icacallbacks.go

```diff
@@ -41,15 +48,110 @@ func (c ICACallbacks) AddICACallback(id string, fn interface{}) icacallbackstype
 }
 
 func (c ICACallbacks) RegisterICACallbacks() icacallbackstypes.ICACallbackHandler {
-	a := c.AddICACallback("samplecallback", ICACallback(SampleCallback))
+	a := c.AddICACallback("delegate", ICACallback(DelegateCallback))
 	return a.(ICACallbacks)
 }
 
 // -----------------------------------
 // ICACallback Handlers
 // -----------------------------------
 
-func SampleCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack []byte, args []byte) error {
-	k.Logger(ctx).Info("samplecallback SUCCESS", "packet", packet, "ack", ack, "args", args)
+// ----------------------------------- Delegate Callback ----------------------------------- //
+func (k Keeper) MarshalDelegateCallbackArgs(ctx sdk.Context, delegateCallback types.DelegateCallback) []byte {
+	out, err := proto.Marshal(&delegateCallback)
+	if err != nil {
+		k.Logger(ctx).Error(fmt.Sprintf("UnmarshalDelegateCallbackArgs %v", err.Error()))
+	}
+	return out
+}
+
+func (k Keeper) UnmarshalDelegateCallbackArgs(ctx sdk.Context, delegateCallback []byte) types.DelegateCallback {
+	unmarshalledDelegateCallback := types.DelegateCallback{}
+	if err := proto.Unmarshal(delegateCallback, &unmarshalledDelegateCallback); err != nil {
+        k.Logger(ctx).Error(fmt.Sprintf("UnmarshalDelegateCallbackArgs %v", err.Error()))
+	}
+	return unmarshalledDelegateCallback
+}
+
+// QUESTION: Would it be cleaner to pass in ack as a bool (success / failure) here?
+func DelegateCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack []byte, args []byte) error {
+	k.Logger(ctx).Info("DelegateCallback executing", "packet", packet, "ack", ack, "args", args)
+	// deserialize the ack
+	txMsgData, err := k.GetTxMsgData(ctx, ack)
+	if err != nil {
+		// ack failed, handle here
+		return err
+	}
+	// do we need txMsgData?
+	_ = txMsgData
+
+	// deserialize the args
+	delegateCallback := k.UnmarshalDelegateCallbackArgs(ctx, args)
```

Comment:
Should we have UnmarshalDelegateCallbackArgs throw an error if it can't unmarshal? Otherwise it will return an empty DelegateCallback{} and I think we'll get a null pointer exception below

Reply:
good call, done